### DISCUSSION TITLE
🐞 OTEL-related environment variables deteriorate overall experience

### DISCUSSION THREAD

**Discussion by exlee** [dagger/dagger#12201] [Help]
### What is the issue?

Hi,

I've been evaluating Dagger and I found performance to be abysmal and TUI exposing buggy and erratic behavior, e.g. (runtimes on M2 Max machine).

### Context

- `dagger functions` took 20-25s (after aggressive `exclude` and filtering in module) to finish
- both TUI and `--progress=plain` tended to hang indefinitely
- TUI often showed weird traces, omitted results or heavily delayed result output
- simple test flow took 6 minutes and didn't finish where raw Docker runtimes for same code took less than 1 minute to execute

The only error emitted during was:
```
ERR failed to emit telemetry error="traces export: context deadline exceeded: rpc error: code = Unavailable desc = name resolver error: produced zero addresses"
```

Because Dagger seems to be a perfect fit for my use case I tried to debug issues I was having. Lack of evidence pointed to issue stemming from my machine which was confirmed by running Dagger inside clean VM.

### Found Cause

The cause was OpenTelemetry ENV variables (see list in collapsed session). Unsetting all resolved the issue completely:

- runtimes dropped below raw Docker baselines
- no more issues
- TUI provided immediate feedback
- `dagger functions` runtimes < 0.1s

### Resolution

In my opinion course of action for this would be:
- Listing issue in Known Issues in Documentation (for visibility)
- Check for ENV variables in CLI and issue a warning
- Make Dagger runner ignore OTEL env variables for task/runner configuration

The last part would be preferred because OTEL variables might be used for tested tasks (the reason why I have those set).

### List of OTEL related variables in my env

- `OTEL_EXPORTER_LOGS_ENABLED`
- `OTEL_TRACES_EXPORTER`
- `OTEL_EXPORTER_OTLP_ENDPOINT`
- `OTEL_BSP_MAX_QUEUE_SIZE`
- `OTEL_EXPORTER_OTLP_INSECURE`
- `OTEL_LOG_LEVEL`
- `OTEL_EXPORTER_OTLP_PROTOCOL`
- `OTEL_TRACES_SAMPLER`
- `OTEL_SERVICE_NAME`

---

I'll consider contributing, but cannot provide any guarantees because of time constraints.

Thank you all for the great work you're doing!  

### Dagger version

dagger v0.13.3 (registry.dagger.io/engine:v0.13.3) darwin/arm64

### Steps to reproduce

- Set OTEL Variables in env prior to runtime
- run `dagger functions` (might require 2-3 runs for errors to manifest, but it is quite consistent)

### Log output

ERR failed to emit telemetry error="traces export: context deadline exceeded: rpc error: code = Unavailable desc = name resolver error: produced zero addresses"

**Comment by vito** [comment]
I'm confused - is the `OTEL_` configuration invalid, i.e. pointing to an unreachable service? The `dagger` CLI very much intentionally respects `OTEL_` env vars and expects them to be a functional configuration. We use it all the time for troubleshooting deeper engine behavior.

**Comment by exlee** [comment]
> OTEL_ configuration invalid, i.e. pointing to an unreachable service?

Using above definition yes, it's invalid for Dagger.

ENV variables are targetting developed application that has a network configuration allowing it to reach OTEL, Dagger Engine isn't configured to do so.

As mentioned above, for me the main issue is lack of information when running functions that Dagger not being able to reach `OTEL_` configured hosts will impact on its performance.
